# Úkol č. 2 - regrese

* Termíny jsou uvedeny na [courses.fit.cvut.cz](https://courses.fit.cvut.cz/BI-ML1/homeworks/index.html).
* Pokud odevzdáte úkol po prvním termínu ale před nejzazším termínem, budete penalizování -12 body, pozdější odevzdání je bez bodu.
* V rámci tohoto úkolu se musíte vypořádat s regresní úlohou, s příznaky různých typů a s chybějícími hodnotami.
* Před tím, než na nich postavíte predikční model, je třeba je nějakým způsobem převést do číselné reprezentace.
    
> **Úkoly jsou zadány tak, aby Vám daly prostor pro invenci. Vymyslet _jak přesně_ budete úkol řešit, je důležitou součástí zadání a originalita či nápaditost bude také hodnocena!**

Využívejte buňky typu `Markdown` k vysvětlování Vašeho postupu. Za nepřehlednost budeme strhávat body.

## Zdroj dat

Budeme se zabývat predikcí délky dožití v různých zemích a letech.
K dispozici máte trénovací data v souboru `data.csv` a data na vyhodnocení v souboru `evaluation.csv`.

#### Seznam příznaků:

* Year - Rok
* Status - Status rozvinuté nebo rozvojové země
* Life expectancy - Délka dožití v letech - **cílová proměnná, kterou budete predikovat**
* Adult Mortality - Úmrtnost dospělých bez ohledu na pohlaví (pravděpodobnost, že osoby, které dosáhly věku 15 let, zemřou před dosažením věku 60 let (uvedeno na 1 000 osob)).
* infant deaths - počet zemřelých kojenců na 1000 obyvatel
* Alcohol - Alkohol, zaznamenaná spotřeba na obyvatele (15+) (v litrech čistého alkoholu)
* percentage expenditure - Výdaje na zdravotnictví v procentech hrubého domácího produktu na obyvatele (%)
* Hepatitis B - pokrytí očkováním proti hepatitidě B (HepB) u dětí ve věku 1 roku (%)
* Measles - Spalničky - počet hlášených případů na 1000 obyvatel
* BMI - průměrný index tělesné hmotnosti celé populace
* under-five deaths - počet úmrtí dětí do pěti let na 1000 obyvatel
* Polio - proočkovanost proti dětské obrně (Pol3) u dětí ve věku 1 roku (%)
* Total expenditure - Výdaje vládních institucí na zdravotnictví jako procento celkových vládních výdajů (%)
* Diphtheria - pokrytí očkováním proti záškrtu, tetanu a černému kašli (DTP3) u jednoletých dětí (%)
* HIV/AIDS - počet úmrtí na 1 000 živě narozených dětí na HIV/AIDS (0-4 roky)
* GDP - hrubý domácí produkt na obyvatele (v USD)
* Population - počet obyvatel země
* thinness 1-19 years - podíl dětí ve věku 10-19 let s indexem tělesné hmotnosti (BMI) menším než 2 směrodatné odchylky pod mediánem (%)
* thinness 5-9 years - podíl dětí ve věku 5-9 let s indexem tělesné hmotnosti (BMI) menším než 2 směrodatné odchylky pod mediánem (%)
* Income composition of resources - Index lidského rozvoje z hlediska příjmového složení zdrojů (index v rozmezí 0 až 1)
* Schooling - počet let školní docházky (roky)


## Pokyny k vypracování

**Body zadání**, za jejichž (poctivé) vypracování získáte **25 bodů**: 
  * V notebooku načtěte data ze souboru `data.csv`. Vhodným způsobem si je rozdělte na podmnožiny, které Vám poslouží pro trénování (trénovací), porovnávání modelů (validační) a následnou predikci výkonnosti finálního modelu (testovací).
    
  * Proveďte základní předzpracování dat:
    * Projděte si jednotlivé příznaky a transformujte je do vhodné podoby pro použití ve vybraném regresním modelu.
    * Nějakým způsobem se vypořádejte s chybějícími hodnotami. _Pozor na metodické chyby!_
    * Můžete využívat i vizualizace. Vše stručně ale náležitě komentujte.
<br /><br />
  * Vytvořte **vlastní implementaci náhodného lesa**. Použijte k tomu níže předpřipravenou kostru.
  
  * Na připravená data postupně aplikujte Vaši předchozí implementaci modelu náhodného lesa, dále jeden z modelů **lineární regrese** nebo **hřebenové regrese**, a alespoň jeden další model podle Vašeho uvážení, přičemž pro každý z těchto modelů přiměřeně:
    * Okomentujte vhodnost daného modelu pro daný typ úlohy.
    * Experimentujte s normalizací (standardizace/min-max), pokud pro daný model očekáváte její příznivý vliv.
    * Vyberte si hlavní hyperparametry k ladění a najděte jejich nejlepší hodnoty (vzhledem k RMSE).
    * Pro model s nejlepšími hodnotami hyperparametrů určete jeho chybu pomocí RMSE a MAE. _Pozor na metodické chyby!_
    * Získané výsledky vždy řádně okomentujte.
<br /><br />
  * Ze všech zkoušených možností v předchozím kroku vyberte finální model a odhadněte, jakou chybu (RMSE) můžete očekávat na nových datech, která jste doposud neměli k dispozici. _Pozor na metodické chyby!_
    
  * Nakonec načtěte vyhodnocovací data ze souboru `evaluation.csv`. Pomocí finálního modelu napočítejte predikce pro tato data. Vytvořte soubor `results.csv`, ve kterém získané predikce uložíte s využitím tří sloupců: **Country**, **Year** a **Life expectancy**. Tento soubor též odevzdejte (uložte do repozitáře vedle notebooku).

  * Ukázka prvních řádků souboru `results.csv`:
  
```
Country,Year,Life expectancy
Peru,2012,71.4
Peru,2013,72.6
...
```


## Poznámky k odevzdání

  * Řiďte se pokyny ze stránky https://courses.fit.cvut.cz/BI-ML1/homeworks/index.html.

## Příprava dat

### Rozdělení dat na testovácí, validační a trénovací

Data reprezentují vývoj v čase podle let. Mohlo by tedy stát za zvážení, zda je namísto náhodného rozdělení nerozdělit podle času. Ze zadání však není zřejmé, jestli budeme predikovat pouze budoucí data, nebo i historická. Vzhledem k obecně rostoucí délce dožití by takovéto rozdělení mohlo vést k výběru modelu, který by systematicky predikoval vyšší hodnoty. To by se zároveň neprojevilo v odhadu přesnosti na testovací množině, která by jakožto poslední v časové řadě měla nejspíše nejvyšší hodnoty predikované proměnné. Náš odhad přesnosti na nových datech by tak mohl být systematicky špatný. Raději tedy zůstaneme u náhodného rozdělení dat v bežném poměru 60:20:20.

In [427]:
import pandas as pd
from sklearn.model_selection import train_test_split

pd.options.display.float_format = '{:.3f}'.format

RANDOM_STATE = 42


all_data = pd.read_csv("data.csv")

x_data = all_data.drop("Life expectancy", axis=1)
y_data = all_data["Life expectancy"]
del all_data

x_train_raw, x_test_raw, y_train, y_test = train_test_split(x_data, y_data, test_size=0.4, random_state=RANDOM_STATE)
x_val_raw, x_test_raw, y_val, y_test = train_test_split(x_test_raw, y_test, test_size=0.5, random_state=RANDOM_STATE)

# Abychom se vyvarovali metodických chyb, další analýza bude probíhat
# pouze s daty z trénovací množiny. Pro stručnost je nazveme data.
train_data = pd.concat((x_train_raw, y_train), axis=1)
data = train_data


### Odebrání sloupců
Na první pohled jsou v datech dvě dvojice velice podobných sloupců: infant deaths s under-five deaths a thinness 1-19 years s thinness 5-9 years. Obě dvojice budou nejspíš skoro linearní závislé, což znamená dvě věci. Zaprvé obsahují prakticky stejnou informaci, tudíž odebráním jednoho ze sloupců z každé dvojice prakticky o nic nepřijdeme a snížíme dimenzi datasetu. Zadruhé se nám skoro lineárně závislé příznaky nehodí pro metodu lineární regrese.

Zdůrazněme, že proměnná `data` v následujících úryvcích kódu vždy obsahuje pouze množinu trénovacích dat.

In [428]:
corr_young_death = data.drop(["Country", "Status"], axis=1).corr().abs()["infant deaths"]["under-five deaths"]
print("korelace infant deaths a under-five deaths: %.3f" % corr_young_death)

corr_thinness = data.drop(["Country", "Status"], axis=1).corr().abs()["thinness  1-19 years"]["thinness 5-9 years"]
print("korelace thinness  1-19 years a thinness 5-9 years: %.3f" % corr_thinness)

korelace infant deaths a under-five deaths: 0.995
korelace thinness  1-19 years a thinness 5-9 years: 0.949


Korelace je velmi vysoká, což ověřuje naši úvahu. Odebereme tedy z každé dvojice sloupec s menším množstvím informace - pro úmrtí odebereme sloupec s kojenci, jelikož jsou obsaženy v úmrtí do pěti let. Ze stejné logiky odebereme sloupce thinness 5-9 years.

In [330]:
COLUMNS_TO_DROP = ["infant deaths", "under-five deaths"]

Dále musíme rozhodnout co se sloupcem Country. Nabývá 183 různých hodnot. To je opravdu hodně na to použít one-hot encoding. Zároveň to ani tak nejsou všechny státy světa, tudíž náš model musí umět predikovat i státy, které nejsou v trénovacích datech. Dále jsou některé jména států poněkud nestandardní (např. The former Yugoslav republic of Macedonia). Nedokážeme tedy ani jednoduše zajistit, že se nám nová státy v nových datech (např. North Macedonia) podaří napárovat na stará data. Z těchto důvodů sloupec Country také odebereme.

In [331]:
COLUMNS_TO_DROP.append("Country")

### Chybějící hodnoty

Vidíme, že chybějících hodnot je v datasetu poměrně hodně. Vzhledem k tomu, že data se z principu výrazně liší na základě státu, mohlo by být poněkud hrubé nahradit chybějící hodnoty mediány přes celý dataset. Budeme tedy nahrazovat hodnoty mediány pro daný stát tam, kde jsme pro stát z trénovacích dat schopni medián spočítat. Tam, kde toho nejsme schopni (protože daný stát v trénovací množině nemáme, nebo ve všech jeho záznamech pro danou hodnotu chybí), použijeme medián ze států, které mají stejný Status (Developing nebo Developed).

Připravíme si dataframe `country_medians`, v kterém si spočítáme medián daného příznaku pro data trénovací množiny dle států. Ten následně budeme používat k doplnění chybějících hodnot. Medián nebudeme počítat pro příznaky Status a Year, pro které to moc nedává smysl a jejichž hodnoty v datasetu nechybí.

In [332]:
grouped = data.drop(["Status", "Year"], axis=1).groupby("Country")
country_medians = grouped.median()
print(country_medians.head(2))

             Adult Mortality  infant deaths  Alcohol  percentage expenditure  \
Country                                                                        
Afghanistan          284.000         78.500    0.010                  21.523   
Albania               18.000          1.000    5.140                 221.843   

             Hepatitis B  Measles    BMI  under-five deaths  Polio  \
Country                                                              
Afghanistan       64.500 1794.000 15.950            108.000 58.000   
Albania           98.000    8.000 53.500              1.000 98.000   

             Total expenditure  Diphtheria  HIV/AIDS      GDP  Population  \
Country                                                                     
Afghanistan              8.265      62.500     0.100  322.962 2806299.000   
Albania                  6.000      98.000     0.100 2416.588  291321.000   

             thinness  1-19 years  thinness 5-9 years  \
Country                         

Následně připravíme mediány přes všechny rozvojové a rozvinuté země.

In [333]:
developing_medians = data[data["Status"] == "Developing"].drop(["Country", "Status", "Year"], axis=1).median()
developed_medians = data[data["Status"] == "Developed"].drop(["Country", "Status", "Year"], axis=1).median()

# status_medians na indexu 0 obsahuje data pro rozvojove zeme a na indexu 1 pro rozvinute
status_medians = pd.concat([developing_medians, developed_medians], axis=1).transpose()
print(status_medians)

# napr medianova hodnota Schooling v rozvinutych zemich je:
print("Mediánová školní docházka v rozvinutých zemích: %.3f" % status_medians.iloc[1]["Schooling"])

   Adult Mortality  infant deaths  Alcohol  percentage expenditure  \
0          161.000          6.000    2.830                  49.589   
1           72.000          0.000   10.400                1127.010   

   Hepatitis B  Measles    BMI  under-five deaths  Polio  Total expenditure  \
0       91.000   20.000 35.000              8.000 89.000              5.410   
1       95.000   15.000 57.150              0.000 96.000              7.980   

   Diphtheria  HIV/AIDS       GDP  Population  thinness  1-19 years  \
0      89.000     0.100  1242.954 1381599.000                 4.800   
1      96.000     0.100 14961.568 1628571.500                 1.000   

   thinness 5-9 years  Income composition of resources  Schooling  \
0               4.800                            0.637     11.700   
1               1.000                            0.865     15.900   

   Life expectancy  
0           69.300  
1           79.200  
Mediánová školní docházka v rozvinutých zemích: 15.900


Metoda pro nahrazení chybějích hodnot bude následující:

In [334]:
def fillna_data(data):
    """Nahradi chybejici hodnoty v datech"""

    def fillna_row(row: pd.Series):
        """Nahradi chybejici hodnoty v danem radku z dat"""
        country = row["Country"]
        # pokud mame data pro tento stat, naplnime jejich median
        if country in country_medians:
            row.fillna(country_medians.loc[country], inplace=True)
        # co se nepovedlo nahradit pomoci statu (nektere maji treba jen nan),
        # to naplnime pomoci medianu zemi se stejnym statusem
        developed = int(row["Developed"])
        row.fillna(status_medians.loc[developed], inplace=True)
        return row

    return data.apply(fillna_row, axis=1)

### Převod Statusu na binární příznak
Na konec se vypořádáme s řetězcovým příznakem Status. Ten může nabývat pouze dvou hodnot - "Developing" a "Developed". Převedeme ho tedy na binární příznak Developed, který bude 0 pro "Developing" a 1 pro "Developed".

Výsledná metoda pro předzpracování dat je tedy následující:

In [335]:
def preprocess_data(data: pd.DataFrame):
    # Prevod Statusu na binarni priznak
    data["Developed"] = data["Status"] == "Developed"
    data.drop("Status", axis=1, inplace=True)

    # Doplneni chybejicich hodnot metodou odvozenou vyse
    data = fillna_data(data)

    # Odebrani sloupcu viz diskuze vyse
    data = data.drop(COLUMNS_TO_DROP, axis=1)

    return data

Výsledek aplikování této metody na trénovací data uložíme do souboru `processed_train_data.csv` (jen pro ukázku).

In [336]:
processed_data = preprocess_data(data)
processed_data.to_csv("processed_train_data.csv")

Zkontrolujeme, že jsme nahradili všechny chybějící hodnoty a že je dataset upraven tak, jak jsme si představovali.

In [337]:
print("chybejicich hodnot celkem:", processed_data.isna().sum().sum())
print(processed_data.head(1))

chybejicich hodnot celkem: 0
     Year  Adult Mortality  Alcohol  percentage expenditure  Hepatitis B  \
207  2010           16.000    8.410                 154.963       86.000   

     Measles   BMI  Polio  Total expenditure  Diphtheria  HIV/AIDS      GDP  \
207        0 5.700  9.000              6.170      86.000     0.100 1595.912   

     Population  thinness  1-19 years  thinness 5-9 years  \
207 1381599.000                 3.800               3.800   

     Income composition of resources  Schooling  Life expectancy  Developed  
207                            0.781     15.800           74.700      False  


Vše vypadá správně, aplikujeme tedy zpracování dat i na validační a testovací data. (`x_train` zde získáme pomocí aplikování metody na původní `x_train_raw`, ale stejně tak by stačilo dropnout sloupec s vysvětlovanou proměnnou z `processed_data`.)

In [338]:
x_train = preprocess_data(x_train_raw)
x_val = preprocess_data(x_val_raw)
x_test = preprocess_data(x_test_raw)

assert x_train.equals(processed_data.drop("Life expectancy", axis=1))

## Trénování modelů

### Náhodný les

Náhodný les má oproti stromu tu výhodu, že je jeho rozhodování více diverzifikované díky různým stromům trénovaným na různých částech dat. Pro regresi tak může průměr odhadů všech stromů být stabilnější a přesnější než odhad jednoho samotného stromu.

Nejdříve dle zadání musíme naimplmenetovat třídu reprezentující model náhodného lesa.

In [368]:
from typing import List

import numpy as np
import pandas as pd
from sklearn.tree import DecisionTreeRegressor


########################################################
# Předpřipravená kostra modelu náhodného lesa
class CustomRandomForest:
    """
    Třída Vašeho modelu
    Bude se jednat o model náhodného lesa, kde podmodely tvoří rozhodovací stromy pro regresi.
    Pro podmodely můžete použít implementaci DecisionTreeRegressor ze sklearn.
    """

    def __init__(self, n_estimators, max_samples, max_depth, **kwargs):
        """
        Konstruktor modelu
        Základní hyperparametery:
            n_estimators - počet podmodelů - rozhodovacích stromů.
            max_samples - tento parametr označuje relativní počet bodů (tj. číslo mezi 0 a 1), které budou pro každý
                          podmodel rozhodovacího stromu náhodně vybrány z trénovací množiny (bootstrap) a použity k jeho trénování.
            max_depth - maximální hloubka každého z podmodelů rozhodovacího stromu.
            kwargs - (volitelně) případné další hyperparametry, které pošlete do podmodelů rozhodovacího stromu
        """
        if not isinstance(max_samples, float):
            raise ValueError("max_sample musi byt zlomek")

        kwargs["max_depth"] = max_depth

        self.trees: List[DecisionTreeRegressor] = [
            DecisionTreeRegressor(**kwargs) for _ in range(n_estimators)
        ]
        self.max_samples = max_samples

        # Respektujeme nastaveni random statu i pro nas vyber dat pro stromy
        random_seed = kwargs.get("random_state", None)
        self.random: np.random.Generator = np.random.default_rng(random_seed)

    def fit(self, X: pd.DataFrame, y: pd.DataFrame):
        """
        Natrénování modelu. Trénovací data jsou v argumentech X a y.
        Pro trénování podmodelů používejte bootstraping a velikost samplovaného vzorku vezměte z hyperparametru max_samples
        """
        for tree in self.trees:
            X_sample, y_sample = self._get_bootstrap_sample(X, y)
            tree.fit(X_sample, y_sample)

    def predict(self, X: pd.DataFrame):
        """
        Predikce y v zadaných bodech X
        """
        ypredicted = np.zeros((X.shape[0],))
        for tree in self.trees:
            # secteme odhady vsech stromu
            ypredicted += tree.predict(X)

        # vezmeme prumer odhadu jednotlivych stromu
        ypredicted /= len(self.trees)
        return ypredicted

    def _get_bootstrap_sample(self, X: pd.DataFrame, y: pd.DataFrame):
        """
        Vybere z dat bootstrapem data pro trenovani jednotlivych stromu
        """
        row_count = X.shape[0]
        sample_size = round(row_count * self.max_samples)
        sample_indices = self.random.choice(row_count, sample_size, replace=True)
        X_sample = X.iloc[sample_indices]
        y_sample = y.iloc[sample_indices]
        return (X_sample, y_sample)

Následně les nátrenujeme na testovacích datech. Ladit budeme hyperparametry `n_estimators`, `max_depth` a `max_samples`, tedy počet stromů v lese, maximální hloubka jednotlivých stromů a velikost vzorku dat, který se má použít pro trénování jednotlivých stromů.

In [340]:
from sklearn.model_selection import ParameterGrid
from sklearn import metrics

param_dict = {
    "n_estimators": range(2, 30, 2),
    "max_depth": range(1, 15),
    "max_samples": [0.3, 0.5, 0.7, 0.9, 1.0],
}

param_grid = ParameterGrid(param_dict)

print(f"Celkem mame {len(param_grid)} kombinaci parametru")

forest_results = []
for params in param_grid:
    model = CustomRandomForest(**params, random_state=RANDOM_STATE)
    model.fit(x_train, y_train)
    rmse = metrics.mean_squared_error(y_val, model.predict(x_val), squared=False)
    forest_results.append((rmse, params, model))

forest_results.sort(key=lambda x: x[0])

Celkem mame 980 kombinaci parametru


In [341]:
best_forest_result = forest_results[0]
best_forest_rmse, best_forest_params, best_forest_model = best_forest_result

print(f"RMSE={best_forest_rmse:.3f}")
best_forest_mae = metrics.mean_absolute_error(y_val, best_forest_model.predict(x_val))
print(f"MAE={best_forest_mae:.3f}")
print("Best forest params:", best_forest_params)

RMSE=2.251
MAE=1.431
Best forest params: {'max_depth': 14, 'max_samples': 1.0, 'n_estimators': 16}


Les s nejmenším RMSE na validační množině je les složený z 16 stromů maximálních hloubek 14, který pro každý strom používá pro trénování stejně datových řádků, jako je v celém datasetu. Na validační množině je jeho RMSE 2.251 a MAE 1.431.

### Lineární regrese

Linearní regrese předpokládá, že je vysvětlovaná proměnná lineárně závislá na příznacích. To v tomto případě nemusí být špatný předpoklad, lineární regrese je tedy na úlohu poměrně vhodná. Problémem pro ni jsou lineárně závislé, nebo skoro lineárně závislé příznaky, což jsme vyřešili již obecně při přípravě dat. Pro lineární regresi, narozdíl od hřebenové, neovlivní přesnost modelu normalizace příznaků, nemusíme se jí tedy zabývat. Pro jistotu, ověření a ze zvědavosti ji však vyzkoušíme. Jinak totiž stejně pro lineární regresi nemáme žádné hyperparametry k ladění.

In [424]:
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler, MinMaxScaler


standard_scaler = StandardScaler()
x_train_standard_scaled = pd.DataFrame(
    standard_scaler.fit_transform(x_train), index=x_train.index, columns=x_train.columns
)
x_val_standard_scaled = pd.DataFrame(
    standard_scaler.transform(x_val), index=x_val.index, columns=x_val.columns
)
x_test_standard_scaled = pd.DataFrame(
    standard_scaler.transform(x_test), index=x_test.index, columns=x_test.columns
)

min_max_scaler = MinMaxScaler()
x_train_min_max_scaled = pd.DataFrame(
    min_max_scaler.fit_transform(x_train), index=x_train.index, columns=x_train.columns
)
x_val_min_max_scaled = pd.DataFrame(
    min_max_scaler.transform(x_val), index=x_val.index, columns=x_val.columns
)
x_test_min_max_scaled = pd.DataFrame(
    min_max_scaler.transform(x_test), index=x_test.index, columns=x_test.columns
)

scales = {
    "None": (x_train, x_val),
    "Standard": (x_train_standard_scaled, x_val_standard_scaled),
    "MinMax": (x_train_min_max_scaled, x_val_min_max_scaled),
}

linreg_results = []
for scale, (x_train_scaled, x_val_scaled) in scales.items():
    model = LinearRegression()
    model.fit(x_train_scaled, y_train)
    rmse = metrics.mean_squared_error(y_val, model.predict(x_val_scaled), squared=False)
    linreg_results.append((rmse, scale, model))

linreg_results.sort(key=lambda x: x[0])

In [425]:
# jen pro hezky vypis
longest_scale_name = len(max(linreg_results, key=lambda x: len(x[1]))[1])
for rmse, scale, _ in linreg_results:
    print(f"normalizace: {scale}{(longest_scale_name - len(scale)) * ' '} - RMSE: {rmse}")

normalizace: None     - RMSE: 4.532494602863496
normalizace: Standard - RMSE: 4.5324946028647695
normalizace: MinMax   - RMSE: 4.53249460286477


Vidíme, že RMSE je pro všechny možnosti opravdu prakticky identické. Vybereme tedy model bez normalizace.

In [327]:
linreg_rmse, _, linreg_model = linreg_results[0]

print(f"RMSE={linreg_rmse:.3f}")
linreg_mae = metrics.mean_absolute_error(y_val, linreg_model.predict(x_val))
print(f"MAE={linreg_mae:.3f}")

RMSE=4.532
MAE=3.318


Lineární regrese má na validační množině RMSE 4.532 a MAE 3.318. Je tedy horší než náhodný les.

### AdaBoost

AdaBoost je šikovný v tom, že je schopne se postupně zaměřit na datové body, které se mu zprvu nepodařilo dobře odhadnout. Toho využijeme v naší úloze a vyzkoušíme, jestli tento přístup bude efektivnější než přístup náhodného lesu. Pro AdaBoost (složený ze stromů potažmo lesů) stejně jako náhodný les a lineární regresi nemá normalizace příznaků význam, nebudeme se jí tedy zabývat.

#### AdaBoost se stromem

In [360]:
from sklearn.ensemble import AdaBoostRegressor
from sklearn.tree import DecisionTreeRegressor

param_dict = {
    "n_estimators": range(5, 100, 5),
    "learning_rate": (0.5, 0.7, 1.0),
    "max_tree_depth": range(1, 15, 2),
}

param_grid = ParameterGrid(param_dict)

ada_boost_tree_results = []
for params in param_grid:
    max_depth = params.pop("max_tree_depth")
    estimator = DecisionTreeRegressor(max_depth=max_depth)
    model = AdaBoostRegressor(estimator=estimator, **params, random_state=RANDOM_STATE)
    model.fit(x_train, y_train)
    rmse = metrics.mean_squared_error(y_val, model.predict(x_val), squared=False)
    params["max_tree_depth"] = max_depth
    ada_boost_tree_results.append((rmse, params, model))

ada_boost_tree_results.sort(key=lambda x: x[0])

In [387]:
best_ada_boost_tree_result = ada_boost_tree_results[0]
best_ada_boost_tree_rmse, best_ada_boost_tree_params, best_ada_boost_tree_model = best_ada_boost_tree_result

print(f"RMSE={best_ada_boost_tree_rmse:.3f}")
best_ada_boost_tree_mae = metrics.mean_absolute_error(y_val, best_ada_boost_tree_model.predict(x_val))
print(f"MAE={best_ada_boost_tree_mae:.3f}")
print("Best ada_boost_tree params:", best_ada_boost_tree_params)

RMSE=2.102
MAE=1.217
Best ada_boost_tree params: {'learning_rate': 1.0, 'n_estimators': 40, 'max_tree_depth': 13}


AdaBoost se stromy má na validční množině  RMSE 2.102, jeho MAE je 1.217. Je tedy zatím naším nejlepším modelem. Nejlepší hyperparametry jsou jedničkový learning rate, 40 stromů a maximální hloubka jednotlivého stromu 13.

#### AdaBoost s lesem

Vyzkoušejme ještě použít AdaBoost s náhodným lesem ze scikitu (k použití vlastní implementace lesu by bohužel bylo potřeba dopsat spoustu technikálií). Jeho trénování bude kvůli dvojtému ensemblu extrémně výpočetně náročné a zdlouhavé. Značně tedy omezíme rozsah hyperparametrů, které budeme ladit a hyperparamatery pro náhodný les budeme volit okolo hodnot, které vycházely nejlépe při použití pouze jednoho lesu.

In [394]:
from sklearn.ensemble import RandomForestRegressor

param_dict = {
    "n_estimators": range(30, 70, 10),
    "learning_rate": (0.9, 1.0,),
    "forest_max_depth": (5, 12, 14, 16),
    "forest_max_samples": (1.0,),
    "forest_n_estimators": (16,),
}

param_grid = ParameterGrid(param_dict)

ada_boost_forest_results = []
for params in param_grid:
    forest_params = {k.replace("forest_", ""): v for k, v in params.items() if k.startswith("forest_")}
    estimator = RandomForestRegressor(**forest_params)
    ada_params = {k: v for k, v in params.items() if not k.startswith("forest_")}
    model = AdaBoostRegressor(estimator=estimator, **ada_params, random_state=RANDOM_STATE)
    model.fit(x_train, y_train)
    rmse = metrics.mean_squared_error(y_val, model.predict(x_val), squared=False)
    ada_boost_forest_results.append((rmse, params, model))

ada_boost_forest_results.sort(key=lambda x: x[0])

In [398]:
best_ada_boost_forest_result = ada_boost_forest_results[0]
best_ada_boost_forest_rmse, best_ada_boost_forest_params, best_ada_boost_forest_model = best_ada_boost_forest_result

print(f"RMSE={best_ada_boost_forest_rmse:.3f}")
best_ada_boost_forest_mae = metrics.mean_absolute_error(y_val, best_ada_boost_forest_model.predict(x_val))
print(f"MAE={best_ada_boost_forest_mae:.3f}")
print("Best ada_boost_forest params:", best_ada_boost_forest_params)

RMSE=2.206
MAE=1.417
Best ada_boost_forest params: {'forest_max_depth': 16, 'forest_max_samples': 1.0, 'forest_n_estimators': 16, 'learning_rate': 1.0, 'n_estimators': 60}


AdaBoost s náhodnými lesy má na validační množině RMSE 2.206 a MAE 1.417. Hyperparametrů jsme ho kvůli náročnosti trénování nenechali vyzkoušet mnoho, těmi nejlepšími z nich byly learning_rate 1.0, počet lesů 60 a hyperparametry lesu: 16 stromů, max_sample 1.0 a maximální hloubka jednotlivého stromu 16. Chyba RMSE AdaBoostu s lesy vyšla lepší než samotného náhodného lesu, ale horší než RMSE AdaBoostu se stromy.

## Finální vyhodnocení

### Volba výsledného modelu

Nejmenší RMSE na validačních datech měl AdaBoost se stromy s learning ratem 1.0, 40 stromy a maximální hloubkou stromu 13. Prohlásíme ho tedy za výsledný model, spočítáme jeho RMSE na testovacích datech a vypočteme predikce k odevzdání.

In [396]:
best_model = best_ada_boost_tree_model

best_model_rmse = metrics.mean_squared_error(y_test, best_model.predict(x_test), squared=False)

print("RMSE na testovacich datech: %.3f" % best_model_rmse)

RMSE na testovacich datech: 1.896


Od našeho finálního modelu můžeme na nových datech očekávat chybu RMSE=1.896, což je překvapivě o cca 10 % lepší než RMSE na validační množině.

### Výpočet predikcí

Na evaluační data aplikujeme na začátku vytvořenou funkci na předzpracování dat a pomocí nejlepšího modelu určíme predikce.

In [412]:
evaluation_data_raw = pd.read_csv("evaluation.csv")
evaluation_data = preprocess_data(evaluation_data_raw)

predictions = best_model.predict(evaluation_data)

res = evaluation_data_raw.loc[:, ["Country", "Year"]]
res["Life expectancy"] = predictions

res.to_csv("results.csv", index=False)